### ISSUE TITLE
Keeping filtered-out data from parquet causes excessive memory usage after 1.15.0

### ISSUE THREAD

**Issue by jackaixin** [pola-rs/polars#20218]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
# prepare some data
import datetime
import polars as pl
import numpy as np

path = '~/tmp/memory_issue_mock_data'

def write_data(date: datetime.date):
    df = pl.DataFrame({
        'date': [date] * 100 * 1000,
        'id': list(range(1, 1001)) * 100,
        'field': [str(i) for i in range(1, 101)] * 1000,
        'value': np.random.rand(100 * 1000)
    }).sort('date', 'id', 'field')

    df.write_parquet(os.path.join(path, f'{date:%Y-%m-%d}.parquet'))

for date in [datetime.date(2014, 1, 1) + datetime.timedelta(days=i) for i in range(3651)]:
    write_data(date)
```

```python
# load data (with different version of polars)
import datetime
import polars as pl

path = '~/tmp/memory_issue_mock_data'
dates = [datetime.date(2018, 1, 1) + datetime.timedelta(days=i) for i in range(1000)]
ids = [1, 2, 3]
fields = ['1', '2', '3']

df = pl.scan_parquet(path).filter(
    pl.col('date').is_in(dates)
    & pl.col('id').is_in(ids)
    & pl.col('field').is_in(fields)
).collect()

# df.estimated_size('mb') -> only about 6MB.
```

### Issue description

With polars versions >=1.15, memory usage when loading data with `scan_parquet` and some filtering is way bigger than what it was for earlier versions.

Note that here I'm only providing a small scale example from my MacBook pro. In my actual use case, the memory usage went up from 4GB to ~40GB, which is a bit too much for my linux box.

### Expected behavior

I tested the code on versions 1.12.0, 1.13.1, 1.14.0, and memory usage was similar, at around 400MB.

However, for versions 1.15, 1.16, 1.17, memory usage shot up a lot:
- 1.17 -> 4.1GB
- 1.16 -> 4.2GB
- 1.15 -> 4.2GB

I suspect there's some caching when scanning parquet files but I'm not sure exactly what's the part that changed since 1.15, and whether there's any parameter that I can tweak to revert to the behavior pre 1.15.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.17.0
Index type:          UInt32
Platform:            macOS-15.1.1-arm64-arm-64bit
Python:              3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]
LTS CPU:             False

----Optional dependencies----
adbc_driver_manager  1.2.0
altair               5.4.1
boto3                <not installed>
cloudpickle          3.0.0
connectorx           0.3.3
deltalake            0.20.1
fastexcel            0.11.6
fsspec               2024.9.0
gevent               24.2.1
google.auth          <not installed>
great_tables         0.12.0
matplotlib           3.8.2
nest_asyncio         1.6.0
numpy                1.26.2
openpyxl             3.1.5
pandas               2.1.4
pyarrow              17.0.0
pydantic             2.9.2
pyiceberg            0.7.1
sqlalchemy           2.0.35
torch                <not installed>
xlsx2csv             0.8.3
xlsxwriter           3.2.0
```

</details>

**Comment by jackaixin** [comment]
@coastalwhite I'm wondering if there's any update on this issue. I just tested with 1.18.0 but the memory issue remains the same.

**Comment by coastalwhite** [comment]
This might, but I am not 100% sure, be a side effect of the `parallel=prefiltered` setting. Could you try with `parallel=columns`?

**Comment by jackaixin** [comment]
@coastalwhite In my example above I was using the default `parallel='auto'` which I thought would pick the 'optimal direction'.  I also tried specifying `parallel='columns'` but got the same result...

**Comment by jackaixin** [comment]
@PrateekArya  I don't think you can to `np.isin(pl.col('date'), dates)`, can you? Could you show me a sample code that uses `np.isin`? Thanks!

**Comment by jackaixin** [comment]
@ritchie46 sorry to bother you but would you have any idea what might be the cause of the issue?

**Comment by coastalwhite** [comment]
Okay, I think I found the culprit. I am pretty sure this issue is a result of the fact that multi-file / hive-partition loading will only early exit after having loaded the row group statistics. That is really a waste, to be honest. It is not really an easy fix, however. I am working #20203 and as part of that effort I will into that as well. When you can test again, I will let you know.

**Comment by coastalwhite** [comment]
I found the actual culprit. It is #19850. This is not really something I want to revert. It will be solved by pushing the actual predicate matching into the parquet reader.

**Comment by jackaixin** [comment]
@coastalwhite Thanks for looking into it! Please let me know when there's a fix I can test.

**Comment by ritchie46** [comment]
We could expose a gc on our arrays, and force a re-materialization if we have less than 5% of original data.

**Comment by PrateekArya** [comment]
> @PrateekArya I don't think you can to `np.isin(pl.col('date'), dates)`, can you?

Hi, 

No, that would probably not work. I did this instead:

`df.filter(np.isin(df['date'], dates))`

I guess it would convert the column to `numpy` array to evaluate `np.isin`. The `filter` just needs a `boolean` mask or truth list of the same size as `df` so it works perfectly here. :)

**Comment by PrateekArya** [comment]
> I found the actual culprit. It is #19850. This is not really something I want to revert. It will be solved by pushing the actual predicate matching into the parquet reader.
> 
> It might be worth it to expose a `POLARS_LOW_MEMORY` option that shrinks the array instead of just slicing in these cases.

I have similar problems with CSV files. Will it also be addressed here? Or is it another issue?

**Comment by jackaixin** [comment]
> > @PrateekArya I don't think you can to `np.isin(pl.col('date'), dates)`, can you?
> 
> 
> 
> Hi, 
> 
> 
> 
> No, that would probably not work. I did this instead:
> 
> 
> 
> `df.filter(np.isin(df['date'], dates))`
> 
> 
> 
> I guess it would convert the column to `numpy` array to evaluate `np.isin`. The `filter` just needs a `boolean` mask or truth list of the same size as `df` so it works perfectly here. :)
> 
> 
> 
> 

Well in that case you can't really use scan_parquet...but have to read the files first before filtering them, which would take up more memory, right?

**Comment by thomas-privat** [comment]
maybe related to [#20851](https://github.com/pola-rs/polars/issues/20851)

**Comment by jackaixin** [comment]
> I found the actual culprit. It is #19850. This is not really something I want to revert. It will be solved by pushing the actual predicate matching into the parquet reader.
> 
> 
> 
> It might be worth it to expose a `POLARS_LOW_MEMORY` option that shrinks the array instead of just slicing in these cases.

Hi @coastalwhite I wonder whether there's any update on this issue and if it's indeed related to #20851 as mentioned above.

**Comment by coastalwhite** [comment]
For Parquet, #20714 added parquet expressions, but they are not enabled by default yet. I did not look into the garbage collection or low memory flag yet.

**Comment by charlieparadox** [comment]
Hi. Seeing exactly the same problem here. Memory consumption after version 1.14 skyrockets for many operations. I tried with the latest version (1.23) and still the same.

**Comment by mlcbby** [comment]
Throwing my hat in the ring here, also running into this issue, nothing to add that hasn't already been mentioned but we have had to put in a few workarounds and bump up cluster memory in order to run our processing. Looking forward to any updtae on this one!

**Comment by coastalwhite** [comment]
This is mostly solved now as the filters are pushed down into the parquet reader. It might still happen for complex predicates, though.

Closing for now.
